scripts/parse-translation.js

```diff
@@ -1,31 +1,113 @@
-var parseKV = require('parse-kv');
-var fs = require('fs');
-var path = require('path');
+const parseKV = require('parse-kv');
+const fs = require('fs');
+const path = require('path');
 
-var basePath = path.join(__dirname, '../');
+const basePath = path.join(__dirname, '../');
+let SPACE_PADDING = '                                                         ';
+var filenameByKey = {};
 
-module.exports = function (shouldParse) {
-  var fileData = readdir(path.join(basePath, 'game/resource/English/'));
+function checkReplaceTokens (allTooltips, key, recursionCheck, source) {
+  key = key.toLowerCase();
+  recursionCheck = recursionCheck || {};
+  if (recursionCheck[key]) {
+    console.error(recursionCheck);
+    throw new Error('Foun d recursive key: ' + key);
+  }
+  recursionCheck[key] = recursionCheck[key] || 0;
+
+  let value = allTooltips[key];
+  let tokens = value.match(/#{[^\s}]+}/g);
```

Comment:
Should this regex be `/${[A-Za-z_]+}/g` to prevent people from doing anything like `#{Hey-There_This[could_be]a_valid_key(In-theory)}`.  That enforces the convention even more so than any number of devs will

Reply:
If they did that then Travis would yell with the error "Hey-There_This[could_be]a_valid_key(In-theory)" was not found.

I do not believe dota actually has any validation rules on keynames. Some randomly even have spaces in them...

You still may be right though, and thanks for the review :)

Reply:
Hmmmm...spaces are a problem if they are valid in keys.  That said, if there are spaces in some keys, the above regex will not find them as spaces are not considered valid in keys.

Reply:
* Your original regex will not find spaces as valid in keys, my apologies...I have been up since 8AM yesterday and it is 2 AM now for me.

Reply:
Yeah, it's basically just "CAPTAINS MODE", the rest all use underscores.

Reply:
One exception makes a possibility.  Either it gets accounted for or it needs to be an override which becomes messy as exceptions grow in number.  Would it be worth building a keys map and just loop through them and do static replacements for expected tokens, check to see if something might be a token and do another pass until it looks translated (Breaking at something like 10 passes to prevent infinite loops.  It could be made to seek for keys and replace them deterministically, but no normalized pattern means the regex is going to be very loose)?